src/errors-new.ts

```diff
@@ -0,0 +1,98 @@
+import {
+  MatcherLocation,
+  MatcherLocationNormalized,
+  RouteLocation,
+  RouteLocationNormalized,
+} from './types'
+
+// Using string enums because error codes are exposed to developers
+// and number enums could collide with other error codes in runtime
+export enum ErrorTypes {
+  NO_ROUTE_MATCH = 'NO_ROUTE_MATCH',
+  INVALID_ROUTE_MATCH = 'INVALID_ROUTE_MATCH',
+  NAVIGATION_GUARD_REDIRECT = 'NAVIGATION_GUARD_REDIRECT',
+  NAVIGATION_ABORTED = 'NAVIGATION_ABORTED',
+  NAVIGATION_CANCELLED = 'NAVIGATION_CANCELLED',
+}
+
+interface RouterError {
+  type: ErrorTypes
+  message: string
+  from?: RouteLocation
+  to?: RouteLocation
+}
+
+const ErrorTypeMessages = {
+  [ErrorTypes.NO_ROUTE_MATCH](
+    location: MatcherLocation,
+    currentLocation?: MatcherLocationNormalized
+  ) {
+    return `No match for\n ${JSON.stringify(location)}${
+      currentLocation
+        ? '\nwhile being at\n' + JSON.stringify(currentLocation)
+        : ''
+    }`
+  },
+  [ErrorTypes.INVALID_ROUTE_MATCH](location: any) {
+    return `Cannot redirect using a relative location:\n${stringifyRoute(
+      location
+    )}\nUse the function redirect and explicitly provide a name`
+  },
+  [ErrorTypes.NAVIGATION_GUARD_REDIRECT](
+    from: RouteLocationNormalized,
+    to: RouteLocation
+  ) {
+    return `Redirected from "${from.fullPath}" to "${stringifyRoute(
+      to
+    )}" via a navigation guard`
+  },
+  [ErrorTypes.NAVIGATION_ABORTED](
+    from: RouteLocationNormalized,
+    to: RouteLocationNormalized
+  ) {
+    return `Navigation aborted from "${from.fullPath}" to "${to.fullPath}" via a navigation guard`
+  },
+  [ErrorTypes.NAVIGATION_CANCELLED](
+    from: RouteLocationNormalized,
+    to: RouteLocationNormalized
+  ) {
+    return `Navigation cancelled from "${from.fullPath}" to "${to.fullPath}" with a new \`push\` or \`replace\``
+  },
+}
+
+export function createRouterError<Type extends ErrorTypes>(
+  type: Type,
+  ...messageArgs: Parameters<typeof ErrorTypeMessages[Type]>
```

Comment:
I think we could have a more generic creation error similar to https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgEoHsCukoFEpTpTIQAekIAJgM7L6HEDeAUMm8mAJ4AOEAXMgDkcQcgA+QgEaDmAX2bNQOeEjoEiAQRLkIVWhmzR6RZC3Yce-ISNbs4A6mCigA5nIVLoKlMagAhbQoaNCwcX1NbNi5eAUFpSORJBydXd2YYTBAEMGB0EGQEKAg4SF8AHlxA3WCDMPUoAD4ACmirXABtQVbBAF0AGmRuOCg4AFtqAQB5UeAwCoGuy1EJAGsITnQYNQYGgEoBSrN2ABsIMBJkAF5kEAgAd22iJt3kOFpcBXMYEybT87XOMhQINhmNqC8IO0Aa9aADNnQelcQSNxlD1jDkHCtq14UMUdQep92EUwJgoPkIGk-hdroViqV6hV6hpmsJBANGK9YgALdaiWS7ZhAA

We could remove the stack entry added by the function call too, similarly to https://github.com/vuejs/vue-router/pull/3047/files#diff-b9b97e885420591ac8b317a611558999R6-R8. Although it's better to check if `stack` exists first. But given that the `stack` property is still not considered stable and the benefit is only removing one stack entry, we don't really need to do it

Reply:
I tried to remove the necessity to specify generic when calling, and ended up with this: https://github.com/vuejs/vue-router-next/blob/ef17a9f75fb1fddb9a0bf946e426079e7c25592b/src/errors-new.ts#L36-L40